distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso

```diff
@@ -1,59 +1,476 @@
+
+            "ab((c)|(d))".find "abc"
+
+       In this case, the group id for "(d)", which is 3, is a valid group id and
+       (Regex.lookup_group 3) will return 3. If the caller tries to get group 3,
+       Match.group will return Nothing.
+
+    lookup_group : Integer | Text -> Integer ! No_Such_Group
+    lookup_group self id =
+        case id of
+            n : Integer -> case (n >= 0 && n < self.internal_regex_object.groupCount) of
+                True -> n
+                False -> Error.throw (No_Such_Group.Error n)
+            name : Text ->
+                # Maps name to number
+                groups = self.internal_regex_object.groups
+
+                n = case groups of
+                    # If Nothing, there are no named groups
+                    Nothing -> Error.throw (No_Such_Group.Error name)
+                    _ -> read_group_map groups name
+                case n of
+                    _ : Integer -> n
+                    Nothing -> Error.throw (No_Such_Group.Error name)
+
+    ## PRIVATE
+
+       Return a lazy iterator over matches against a string.
+
+       Arguments
+       - text: the string to match against.
+    iterator : Text -> Match_Iterator
+    iterator self input = Match_Iterator.new self input
+
+    ## Return the number of groups in the underlying RegexObject.
+       Note, the count includes group 0 (the whole match) as well.
+    group_count : Integer
+    group_count self = self.internal_regex_object.groupCount
+
+    ## Return a vector of all named group names.
+    named_groups : Vector Text
+    named_groups self =
+        map = polyglot_map_to_map self.internal_regex_object.groups
+        map.keys
+
+    ## Return a map from group number to group name. Only includes named groups.
+    group_nums_to_names : Map Integer Text
+    group_nums_to_names self =
+        map = polyglot_map_to_map self.internal_regex_object.groups
+        map.transform k-> v-> [v, k]
+
+    ## Compile the provided `expression` into a regex pattern that can be used for
+       matching.
+
+       Arguments
+       - expression: The text representing the regular expression that you want to
+         compile. Must be non-empty.
+       - case_insensitive: Enables or disables case-insensitive matching. Case
+         insensitive matching behaves as if it normalises the case of all input
+         text before matching on it.
+
+       If an empty regex is used, `compile` throws an Illegal_Argument error.
+
+       ? Why Compile?
+         While many regex engines are able to cache ad-hoc patterns, it is often
+         useful to be able to manually retain a pattern that you have computed. This
+         function exists so you can hold onto the resultant `Regex` object,
+         instead of immediately proceeding to match using it.
+    compile : Text -> Boolean | Nothing -> Regex ! Regex_Syntax_Error | Illegal_Argument
+    compile expression case_insensitive=Nothing =
```

Comment:
I'd put the statics at the top of the type. Constructor method nexts to actual Constructor.

Reply:
Put `compile` at the top; left `escape` because it's not a constructor method.